### ISSUE TITLE
[Central Management] Add K/V metadata

### ISSUE THREAD

**Issue by andrewvc** [elastic/beats#9881]
It's useful to associated K/V data with an enrolled beat to support advanced applications and interfaces.

For the Uptime app specifically we need a way to know which logical location (which is a user-defined string like `us-east-1a`) a given heartbeat instance is in. We need this information to catalog which locations are available for a monitor to run in. For Uptime this translates directly to checkboxes on a 'create monitor page'.

Requirements:

1. Add Support for K/V metadata to the CM implementation in Kibana.
2. Add a GUI/API for changing K/V metadata to Kibana
3. Add a flag on beat enroll, say `-E location=us-east-1a` that would let the user set an initial value here. This is important for our use case because in most cases that value would never change and it is easy to determine it at the time the beat is deployed.

**Comment by ph** [comment]
To make less global and more scoped we should probably prefix the `location` with either `meta` or `metadata`, internally it would be a `map[string]string` or a `map[string]interface`, `-E` is handled by ucfg so the `meta` key will need to be present in the config and we will just make the logic to send the information as part of the enrollement process.

**Comment by andrewvc** [comment]
++ to the meta prefix. I assume that means all default configs should
contain meta? I'd rather avoid people adding that themselves

**Comment by andrewvc** [comment]
I'm actually thinking relying on default configs is a bad idea. Does that
key need to eb in yaml somewhere, or can we make it in memory?

**Comment by ph** [comment]
@andrewvc The problem is yes, if we use `-E` it will need to deal with configuration present in the yaml file. If we want to separate both behavior we might need to add a new flag which is only used by the enroll subcommand, `--meta`

**Comment by andrewvc** [comment]
@ph in that case I think we need to use a new flag.

**Comment by ph** [comment]
Notes from our sync on this issue:

Step 1:

1.  Add the fields to the Config YAML
2. Uses the -E at enroll time to persist them.
3. Send the values when beats starts at checking time.
4. If we plan to allow Kibana to add metadata fields mark them as disabled for now.

Step 2:

- Investigate if we allow synchronization of the fields from kibana to the beats.

**Comment by jalvz** [comment]
This could be useful for APM as well. I didn't fully explore this, but if libbeat provides an API for an enrolled beat to set key,value pairs; we could use that in the UI to show what apm agents are talking to what apm servers

**Comment by mattapperson** [comment]
@jalvz What kind of scale would we need to support for that? I am thinking that might be a very large list of agents. IDK if a K/V meta-data API would be the right use-case for that. It would cause some major perf issues.

**Comment by jalvz** [comment]
No, I think it shouldn't be problematic, should we go down this path (if you were thinking eg RUM agent, we can not show all the instrumented apps in this neither any other practical way, so that is ruled out)

**Comment by mattapperson** [comment]
@jalvz How many agents are you expecting to list in the K/V store?

**Comment by jalvz** [comment]
Worst case 20 or 30. We just need to store some kind of ids to relate server-agents. 
Forgot to mention, but such API would require also removal and querying of k-v pairs and be exposed in libbeat.

**Comment by ph** [comment]
@jalvz can clarify the use case for removal of keys? We initially thought that beats would provide theses values at checking time from configured values in the YAML, if we allow users to remove keys from the ui this open different problematic of synchronization.

**Comment by jalvz** [comment]
No, our use case is not the same as Uptime. Re-reading this now, I understand that there would be a way in Kibana to edit metadata.

What we need instead is that apm-server call a function eg like `SetMetadata(beatId, arbitraryThing`)
This wouldn't be a hooked into configuration whatsoever. 
I imagine then that we can store that data in Kibana so we can show a list of enrolled apm-servers, and underneath each, a list of apm agents talking to them. Like so:

- `apm-server-1`
 ---- `apm-agent-python-1`
 ---- `apm-agent-python-2`
 ---- `apm-agent-java-1`
- `apm-server-2`
 ---- `apm-agent-java-2`

The need for removing metadata comes from the case where apps are no longer instrumented by an agent, we should remove them from the list. But that can also be achieve with some other ways probably.

**Comment by ph** [comment]
> The need for removing metadata comes from the case where apps are no longer instrumented by an agent, we should remove them from the list. But that can also be achieved with some other ways probably.

We have logic in place to detect when a beat has not fetched a configuration for a period of time we could expire the above data using a timeout, or your use case requires a human intervention?

**Comment by mattapperson** [comment]
@ph ummmm that logic is display/request based. It can not be used to serve as a cleanup tool at this time. (though we plan to add that in the future)

**Comment by ph** [comment]
@mattapperson Good point, I was just worried about having data around or having to synchronize between Kibana<->beats.

**Comment by jalvz** [comment]
Ok, lets say the ability for removing is not the most important for 7.1, but please keep it in mind.
I think if we can show a last update next to the agents in a similar way of how it happens now for beats, that would be good enough.

Generally, I imagine this feature as a set of CRUD API operations for Beats to PUT, GET (and eventually DELETE) arbitrary data associated to a beat-Id, backed by a programmatic API in Libbeat. 
It is important for APM that the action can be driven by apm-server, but probably could be useful as well that such metadata can be created/removed from Kibana, should the UI for a beat enable that.

Does this make sense? Am I missing other important considerations?

**Comment by ph** [comment]
@jalvz, when you say that this is important that the information is driven by apm-server, are you implying that information would change at runtime? Because our original discussion and feature definition was in the concept of beats checking in when booting in OR enrolling.

Also does that mean information could be change from an agent connected to apm-server?

**Comment by jalvz** [comment]
Yes, that is correct, the metadata would be added, queried, and removed at runtime.

**Comment by ph** [comment]
@jalvz So would I be correct to say, we could potentially have two kind of metadata.

1. from beats (apm-server, filebeat etc)
2. User defined in the UI.

> queried 

Queried by the beats?

**Comment by ph** [comment]
We will do the following:

- We will not accept nested objects for the metadata, supporting nested structures will complexify the UI adding more validations and UI interaction. We think that having support for key -> value, where value can be a string, a number, or array of string or number should be enough for the majority of use cases.
- Metadata are sends when we enroll or when the beat start.
- Metadata configured by the beats will be locked in the UI.
- Metadata can be added by the user, but it will not get synchronized to the beat.

Implementations steps:

 1. We will release the initial support from the beat.
 2. Display and searchable in the UI
 3. Adding/editing/deleting from the UI.

**Comment by mattapperson** [comment]
In addition. Once we have the UI. Configs set by the beat itself will be read-only in the UI

**Comment by andrewvc** [comment]
@ph that all sounds great!

**Comment by andrewvc** [comment]
@mattapperson I assume, btw, we'll be able to aggregate on all values for the metadata once it's in ES right? Like, if I want to find all values with a given key?

LMK if we should move that convo to a different issue.

**Comment by mattapperson** [comment]
At the moment, no. You could do that client side for smaller result sets up to hundreds of results. If you need to scale beyond that you can open a feature request and I’ll be happy to add support. It’s a nominal enhancement but would probably require a new endpoint...
